rubullet/src/client.rs

```diff
@@ -0,0 +1,3254 @@
+                    &mut sensor_state,
+                ) {
+                    result_list_joint_states.push(sensor_state.into());
+                } else {
+                    return Err(Error::new("getJointState failed (2)."));
+                }
+            }
+            Ok(result_list_joint_states)
+        }
+    }
+    /// calculate_mass_matrix will compute the system inertia for an articulated body given
+    /// its joint positions.
+    /// The composite rigid body algorithm (CBRA) is used to compute the mass matrix.
+    /// # Arguments
+    /// * `body` - the [`BodyId`](`crate::types::BodyId`), as returned by [`load_urdf`](`Self::load_urdf()`) etc.
+    /// * `object_positions` - joint positions for each link
+    /// # Return
+    /// The result is the square mass matrix with dimensions dofCount * dofCount, stored as a
+    /// list of dofCount rows, each row is a list of dofCount mass matrix elements.
+    /// Note that when multidof (spherical) joints are involved, calculate_mass_matrix will use a
+    /// different code path, that is a bit slower.
+    pub fn calculate_mass_matrix(
+        &mut self,
+        body: BodyId,
+        object_positions: &[f64],
+    ) -> Result<DMatrix<f64>, Error> {
+        if !object_positions.is_empty() {
+            let joint_positions = object_positions;
+            let flags = 0; // TODO add flags
+            unsafe {
+                let command_handle = ffi::b3CalculateMassMatrixCommandInit(
+                    self.handle.as_ptr(),
+                    body.0,
+                    joint_positions.as_ptr(),
+                    joint_positions.len() as i32,
+                );
+                ffi::b3CalculateMassMatrixSetFlags(command_handle, flags);
+                let status_handle =
+                    ffi::b3SubmitClientCommandAndWaitStatus(self.handle.as_ptr(), command_handle);
+                let status_type = ffi::b3GetStatusType(status_handle);
+                if status_type == CMD_CALCULATED_MASS_MATRIX_COMPLETED as i32 {
+                    let mut dof_count = 0;
+                    ffi::b3GetStatusMassMatrix(
+                        self.handle.as_ptr(),
+                        status_handle,
+                        &mut dof_count,
+                        std::ptr::null_mut(),
+                    );
+                    if dof_count != 0 {
+                        let mut mass_vec = vec![0.; (dof_count * dof_count) as usize];
+                        ffi::b3GetStatusMassMatrix(
+                            self.handle.as_ptr(),
+                            status_handle,
+                            &mut dof_count,
+                            mass_vec.as_mut_slice().as_mut_ptr(),
+                        );
+                        let mass_mat = DMatrix::<f64>::from_row_slice(
+                            dof_count as usize,
+                            dof_count as usize,
+                            mass_vec.as_slice(),
+                        );
```

Comment:
Changed it. Did you know that from_vec takes a column vector and that it only works in this case because mass matrices are symmetric? First, I wanted to to the same for the jacobian but the tests kept failing :see_no_evil:

Reply:
In those cases, you can always `DMatrix::from_vec` and then `Matrix::transpose_mut`.